### PR TITLE
Implement DeRegister for graceful party shutdown

### LINKED ISSUE #5: Deregister
De-register party from aggregator and graceful shutdown.

- [ ] Implement De-register message dataclass.
- [ ] Send de-register request to aggregator when a party is shutting down. 
- [ ] Handle de-register message on the aggregator to remove party from client set.

### PR DESCRIPTION
**Pull request by Ceres445**
Closes #5 

> De-register party from aggregator and graceful shutdown.
> 
> * [x]  Implement De-register message dataclass.
> * [x]  Send de-register request to aggregator when a party is shutting down.
> * [x]  Handle de-register message on the aggregator to remove party from client set.

### CODE REVIEW

**Comment by Ceres445** [comment]
Doesn't work on [docker](https://github.com/encode/uvicorn/issues/1215) containers .

I'm checking if there is a workaround

**Comment by Ceres445** [comment]
Works for docker containers now

File: tinyfl/aggregator/__init__.py [hunk 1]
```diff
@@ -110,10 +110,16 @@ async def handle(req: Request, background_tasks: BackgroundTasks):
         case Register(url=url):
             with client_lock:
                 clients.add(url)
+            logger.info("Client registered")
```
Comment (eigengravy) [inline tinyfl/aggregator/__init__.py:113]: log the client's url as well
```suggestion
            logger.info(f"Client {url} registered")
```

File: tinyfl/aggregator/__init__.py [hunk 2]
```diff
@@ -110,10 +110,16 @@ async def handle(req: Request, background_tasks: BackgroundTasks):
...
             return {"success": True, "message": "Registered"}
         case SubmitWeights(round=round, weights=weights):
             background_tasks.add_task(collect_weights, copy.deepcopy(weights))
             return {"success": True, "message": "Weights submitted"}
+        case DeRegister(url=url):
+            with client_lock:
+                clients.remove(url)
+            logger.info("Client de-registered")
```
Comment (eigengravy) [inline tinyfl/aggregator/__init__.py:121]: same as above
```suggestion
            logger.info(f"Client {url} de-registered")
```

**Review by eigengravy — CHANGES_REQUESTED** [review]

**Review by eigengravy — APPROVED** [review]
lgtm

**Pull request merged**